RFC/0001-graph-display.md

```diff
@@ -0,0 +1,241 @@
+- It is only possible to examine one module at a time.
+- Values shared between nodes are contained in a keyword system that
+  can be both opaque (hard to determine keyword values) and brittle
+  (long but similar keyword names that must be typed exactly).
+- It is difficult for users to create complicated layers without
+  outside help.
+- There is no method of gaining a high level overview of the
+  activities of a layer.  Each module much be read in sequence and a
+  mental modal of the entire, possibly non-linear, process must be
+  constructed from this.
+  
+A graphical display of a connected nodes provides an alternative that
+ameliorates most of these issues.
+
+- All modules can be seen at once with the relations between modules
+  explicitly declared by the connection.
+- All modules and their values are present on the workspace at the same time.
+- Values are shared by drawing connections between modules instead of
+  variables from a global namespace.
+- The graphical interface provides an intuitive framework for treating
+  the modules as individual building blocks.
+- One glance at the workflow gives an overview of all of the modules
+  and their relations
+
+
+## 3 Proposed Implementation
+
+The base of the implementation is an `AbstractGraphModel` class.
+
+```mermaid
+classDiagram
+  class AbstractGraphModel {
+      +nodes() : AbstractListModel~NodeInterface~
+      +edges() : AbstractListModel~EdgeInterface~
+      #createNode(string)
+      #deleteNode(int)
+      #createEdge(int, int, int, int)
+      #deleteEdge(int)
+  }
+  <<Abstract>> AbstractGraphModel
+  class NodeInterface {
+    + x : float
+    + y : float
+    + name: String
+    + icon: Icon
+    + type: String
+  }
+  <<Interface>> NodeInterface
+  class EdgeInterface {
+    + startHandle : Point2D
+    + start : Point2D
+    + end : Point2D
+    + endHandle : Point2D
+    + style : int
+  }
+  <<Interface>> EdgeInterface
+  AbstractGraphModel *-- NodeInterface
+  AbstractGraphModel *-- EdgeInterface
+```
+
+### 3.1 AbstractGraphModel
+
+The `AbstractGraphModel` defines the interface to our graph.  The
+internal representation of the graph is both beyond the scope of this
+document[^1] and independent of the behaviour of the library.  All
+that matters is that the class which represents the graph contains two
+Qt properties.  The first property is `nodes`, which returns an object
+that is a subclass of `AbstractListModel` and each node returned
+implements the `NodeInterface` in the Qt properties.  The other
+property, `edges`, is similar, except that the mandatory interface is
+the `EdgeInterface`.
+
+The `AbstractGraphModel` also requires four slots for accepting
+signals from the interface.  The `createNode` slot takes a string that
+corresponds to the `type` property of the Node Interface and appends a
+new Node of the correct type.  Correspondingly, `deleteNode` takes the
+index of a node and removes that node from the model.  In the same
```

Comment:
I think it would be wise to continue demanding unique names for nodes / modules, so we don't need to access by `index` and could use the object name instead.